### PR TITLE
New facts module for AWS VPC Peering

### PR DESCRIPTION
**Pull request by willthames**
##### SUMMARY
Retrieves AWS VPC Peering details using AWS methods. Requires Boto3.
- Gets various details related to AWS VPC Peers
  Filters can be found http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeVpcPeeringConnections.html

##### ISSUE TYPE
Feature Pull Request

##### COMPONENT NAME
ec2_vpc_peering_facts module

##### ANSIBLE VERSION
2.3

### CODE REVIEW

**Comment by willthames** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/1431 by willthames (not original author)

**Comment by zyv** [comment]
Tested on Ansible 2.2, works nicely for me, would be awesome if this could be included upstream! Need such a module to construct routing tables by querying VPC peerings.

**Comment by michaeljs1990** [comment]
ok, this is a little ridiculous. I'm all for helping out an open source project but this is generating a huge amount of spam now. I don't see an option to have the bot remove me from the list.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_peering_facts.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+    try:
+        region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+        ec2 = boto3_conn(module, conn_type='client', resource='ec2', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.NoCredentialsError as e:
+        module.fail_json(msg=str(e))
+
+    results = get_vpc_peers(ec2, module)
+
+    module.exit_json(result=results)
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_peering_facts.py]: For results, it looks like you're returning the camelcase output from boto3. Can you please use `camel_dict_to_snake_dict` to normalize the returned values?

**Review by ryansb — CHANGES_REQUESTED** [review]
Otherwise LGTM, just needs to return normalized snake case variables.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_peering_facts.py (hunk 1 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_peering_facts.py]: Done

**Comment by willthames** [comment]
!needs_revision

**Comment by mattclay** [comment]
Closing and re-opening to trigger CI.

**Comment by mattclay** [comment]
CI failure due to module validation failures:

```
2017-02-03 01:52:35 ============================================================================
2017-02-03 01:52:35 lib/ansible/modules/cloud/amazon/ec2_vpc_peering_facts.py
2017-02-03 01:52:35 ============================================================================
2017-02-03 01:52:35 ERROR: No ANSIBLE_METADATA provided
2017-02-03 01:52:35 ERROR: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 17
2017-02-03 01:52:35 ERROR: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 20
2017-02-03 01:52:35 ERROR: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 21
2017-02-03 01:52:35 ERROR: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 22
2017-02-03 01:52:35 ERROR: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 23
```

File: lib/ansible/modules/cloud/amazon/ec2_vpc_peering_facts.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    try:
+        region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+        ec2 = boto3_conn(module, conn_type='client', resource='ec2', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.NoCredentialsError as e:
+        module.fail_json(msg=str(e))
+
+    results = [camel_dict_to_snake_dict(peer) for peer in get_vpc_peers(ec2, module)]
+
+    module.exit_json(result=results)
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_peering_facts.py:144]: For facts modules it's standard to return the `ansible_facts` key as in [cloudfront_facts](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/cloudfront_facts.py#L518) and others. Can you make the results available as facts under a key such as `vpc_peering` or `ec2_vpc_peering`?

**Review by ryansb — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
It seems that ship has long since sailed - is there any actual benefit to that 'standard'?

Adhering modules:
```
cloudformation_facts.py
cloudfront_facts.py
ec2_elb.py
ec2_elb_lb.py
ec2_facts.py
ecs_service.py
ecs_service_facts.py
efs_facts.py
lambda_facts.py
```

Non-adhering modules
```
ec2_asg_facts.py
ec2_elb_facts.py
ec2_eni_facts.py
ec2_group_facts.py
ec2_lc_facts.py
ec2_remote_facts.py
ec2_snapshot_facts.py
ec2_vol_facts.py
ec2_vpc_dhcp_options_facts.py
ec2_vpc_igw_facts.py
ec2_vpc_nacl_facts.py
ec2_vpc_nat_gateway_facts.py
ec2_vpc_net_facts.py
ec2_vpc_route_table_facts.py
ec2_vpc_subnet_facts.py
ec2_vpc_vgw_facts.py
iam_mfa_device_facts.py
iam_server_certificate_facts.py
route53_facts.py
```

**Comment by ryansb** [comment]
Part of the reason they don't all store in that way is that the `ansible_facts` argument was added after some of those modules were created. I think there is a benefit to setting them as facts in the module, but if community members don't agree, I'm not wedded to the idea. I'll bring it up at the Core meeting today.

**Comment by dagwieers** [comment]
@ryansb ansible_facts was there from the very beginning, it predates all these modules :-) I think it is just an oversight of the developer who probably was more used to register stuff. However these are 2 very distinct mechanisms.

**Comment by ryansb** [comment]
Based on discussion in the meeting, we reached a consensus that ansible_facts  doesn't have to be included in new modules. Clearing my review.

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
shipit

**Comment by zyv** [comment]
Just tested the new version with Ansible 2.3 and it works fine for me, just had to adjust playbooks re. snake case / camel case. Looks good to me, a very useful module! Not sure what else I could do to get this merged eventually, too bad it didn't make it into 2.3...

shipit

**Review by s-hertel — APPROVED** [review]

**Pull request merged**